Bug: Integer parsing exception during constant resolution

## Describe the bug
We attempt to stick all strings consisting of only digits into an `IntegerType` which can throw an exception if the numbers is too large (or too small).

## Steps to reproduce
Steps to reproduce the behavior:
1. Run the unit test `test fhir path resolver multiple values` with a string field consisting of only numbers greater than `Integer.MAX`

## Expected behavior
Attempt to parse a strings of only digits into an integer but fallback to a string if an exception is thrown.

## Logs
```
Exception while executing function: Functions.translate-fhir Result: Failure
Exception: NumberFormatException: For input string: "6000352053"
Stack: java.lang.reflect.InvocationTargetException
	at jdk.internal.reflect.GeneratedMethodAccessor1013.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.microsoft.azure.functions.worker.broker.JavaMethodInvokeInfo.invoke(JavaMethodInvokeInfo.java:22)
	at com.microsoft.azure.functions.worker.broker.EnhancedJavaMethodExecutorImpl.execute(EnhancedJavaMethodExecutorImpl.java:22)
	at com.microsoft.azure.functions.worker.chain.FunctionExecutionMiddleware.invoke(FunctionExecutionMiddleware.java:19)
	at com.microsoft.azure.functions.worker.chain.InvocationChain.doNext(InvocationChain.java:21)
	at com.microsoft.azure.functions.worker.broker.JavaFunctionBroker.invokeMethod(JavaFunctionBroker.java:125)
	at com.microsoft.azure.functions.worker.handler.InvocationRequestHandler.execute(InvocationRequestHandler.java:34)
	at com.microsoft.azure.functions.worker.handler.InvocationRequestHandler.execute(InvocationRequestHandler.java:10)
	at com.microsoft.azure.functions.worker.handler.MessageHandler.handle(MessageHandler.java:44)
	at com.microsoft.azure.functions.worker.JavaWorkerClient$StreamingMessagePeer.lambda$onNext$0(JavaWorkerClient.java:94)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: gov.cdc.prime.router.fhirengine.translation.hl7.schema.ConfigSchemaElementProcessingException: Error encountered while applying: ca-order-observations in CA-receiver-transform to FHIR bundle. 
Error was: Error encountered while applying: ca-specimen in ca-order-observation to FHIR bundle. 
Error was: Error encountered while applying: ca-specimen-organization-identifier in ca-specimen to FHIR bundle. 
Error was: Error encountered while applying: organization-identifier in ../../../common/organization-identifier to FHIR bundle. 
Error was: java.lang.NumberFormatException: For input string: "6000352053"
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processSchema(FhirToHl7Converter.kt:123)
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processSchema$default(FhirToHl7Converter.kt:101)
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.convert(FhirToHl7Converter.kt:91)
	at gov.cdc.prime.router.fhirengine.engine.FHIRTranslator.getHL7MessageFromBundle$prime_router(FHIRTranslator.kt:188)
	at gov.cdc.prime.router.fhirengine.engine.FHIRTranslator.getByteArrayFromBundle$prime_router(FHIRTranslator.kt:164)
	at gov.cdc.prime.router.fhirengine.engine.FHIRTranslator.doWork(FHIRTranslator.kt:75)
	at gov.cdc.prime.router.fhirengine.engine.FHIREngine.run(FHIREngine.kt:173)
	at gov.cdc.prime.router.fhirengine.azure.FHIRFunctions$runFhirEngine$newMessages$1.invoke(FHIRFunctions.kt:150)
	at gov.cdc.prime.router.fhirengine.azure.FHIRFunctions$runFhirEngine$newMessages$1.invoke(FHIRFunctions.kt:149)
	at gov.cdc.prime.router.azure.DatabaseAccess.transactReturning$lambda$1(DatabaseAccess.kt:106)
	at org.jooq.impl.DefaultDSLContext.lambda$transactionResult0$3(DefaultDSLContext.java:531)
	at org.jooq.impl.Tools$3$1.block(Tools.java:6068)
	at java.base/java.util.concurrent.ForkJoinPool.unmanagedBlock(ForkJoinPool.java:3463)
	at java.base/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3434)
	at org.jooq.impl.Tools$3.get(Tools.java:6065)
	at org.jooq.impl.DefaultDSLContext.transactionResult0(DefaultDSLContext.java:579)
	at org.jooq.impl.DefaultDSLContext.transactionResult(DefaultDSLContext.java:503)
	at gov.cdc.prime.router.azure.DatabaseAccess.transactReturning(DatabaseAccess.kt:106)
	at gov.cdc.prime.router.fhirengine.azure.FHIRFunctions.runFhirEngine(FHIRFunctions.kt:149)
	at gov.cdc.prime.router.fhirengine.azure.FHIRFunctions.doTranslate(FHIRFunctions.kt:129)
	at gov.cdc.prime.router.fhirengine.azure.FHIRFunctions.doTranslate$default(FHIRFunctions.kt:123)
	at gov.cdc.prime.router.fhirengine.azure.FHIRFunctions.translate(FHIRFunctions.kt:114)
	... 17 more
Caused by: gov.cdc.prime.router.fhirengine.translation.hl7.schema.ConfigSchemaElementProcessingException: Error encountered while applying: ca-specimen in ca-order-observation to FHIR bundle. 
Error was: Error encountered while applying: ca-specimen-organization-identifier in ca-specimen to FHIR bundle. 
Error was: Error encountered while applying: organization-identifier in ../../../common/organization-identifier to FHIR bundle. 
Error was: java.lang.NumberFormatException: For input string: "6000352053"
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processSchema(FhirToHl7Converter.kt:123)
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processElement$prime_router(FhirToHl7Converter.kt:170)
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processSchema(FhirToHl7Converter.kt:121)
	... 38 more
Caused by: gov.cdc.prime.router.fhirengine.translation.hl7.schema.ConfigSchemaElementProcessingException: Error encountered while applying: ca-specimen-organization-identifier in ca-specimen to FHIR bundle. 
Error was: Error encountered while applying: organization-identifier in ../../../common/organization-identifier to FHIR bundle. 
Error was: java.lang.NumberFormatException: For input string: "6000352053"
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processSchema(FhirToHl7Converter.kt:123)
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processElement$prime_router(FhirToHl7Converter.kt:170)
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processSchema(FhirToHl7Converter.kt:121)
	... 40 more
Caused by: gov.cdc.prime.router.fhirengine.translation.hl7.schema.ConfigSchemaElementProcessingException: Error encountered while applying: organization-identifier in ../../../common/organization-identifier to FHIR bundle. 
Error was: java.lang.NumberFormatException: For input string: "6000352053"
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processSchema(FhirToHl7Converter.kt:123)
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processElement$prime_router(FhirToHl7Converter.kt:170)
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processSchema(FhirToHl7Converter.kt:121)
	... 42 more
Caused by: java.lang.IllegalArgumentException: java.lang.NumberFormatException: For input string: "6000352053"
	at org.hl7.fhir.r4.model.IntegerType.parse(IntegerType.java:90)
	at org.hl7.fhir.r4.model.IntegerType.parse(IntegerType.java:39)
	at org.hl7.fhir.r4.model.PrimitiveType.fromStringValue(PrimitiveType.java:108)
	at org.hl7.fhir.r4.model.PrimitiveType.setValueAsString(PrimitiveType.java:164)
	at org.hl7.fhir.r4.model.IntegerType.<init>(IntegerType.java:66)
	at gov.cdc.prime.router.fhirengine.translation.hl7.utils.FhirPathCustomResolver.resolveConstant(ConstantResolver.kt:167)
	at org.hl7.fhir.r4.utils.FHIRPathEngine.resolveConstant(FHIRPathEngine.java:1729)
	at org.hl7.fhir.r4.utils.FHIRPathEngine.resolveConstant(FHIRPathEngine.java:1625)
	at org.hl7.fhir.r4.utils.FHIRPathEngine.execute(FHIRPathEngine.java:1476)
	at org.hl7.fhir.r4.utils.FHIRPathEngine.evaluate(FHIRPathEngine.java:664)
	at gov.cdc.prime.router.fhirengine.translation.hl7.utils.FhirPathUtils.evaluateString(FhirPathUtils.kt:166)
	at gov.cdc.prime.router.fhirengine.translation.hl7.ConfigSchemaProcessor.getValueAsString$prime_router(ConfigSchemaProcessor.kt:34)
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processElement$prime_router(FhirToHl7Converter.kt:181)
	at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processSchema(FhirToHl7Converter.kt:121)
	... 44 more
Caused by: java.lang.NumberFormatException: For input string: "6000352053"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
	at java.base/java.lang.Integer.parseInt(Integer.java:668)
	at java.base/java.lang.Integer.parseInt(Integer.java:786)
	at org.hl7.fhir.r4.model.IntegerType.parse(IntegerType.java:88)
	... 57 more
```

--- a/prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/ConstantResolver.kt
+++ b/prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/ConstantResolver.kt
@@ -17,6 +17,8 @@ import org.hl7.fhir.r4.model.TypeDetails
 import org.hl7.fhir.r4.model.ValueSet
 import org.hl7.fhir.r4.utils.FHIRPathEngine
 import org.hl7.fhir.r4.utils.FHIRPathUtilityClasses.FunctionDetails
+import java.lang.IllegalArgumentException
+import java.lang.NumberFormatException
 
 /**
  * Context used for resolving [constants] and custom FHIR functions. The class is for us to add our customer function
@@ -166,8 +168,10 @@ class FhirPathCustomResolver(private val customFhirFunctions: FhirPathFunctions?
                     if (it is StringType && StringUtils.isNumeric(it.primitiveValue())) {
                         try {
                             IntegerType(it.primitiveValue())
-                        } catch (e: NumberFormatException) {
-                            it // fallback to string; see https://github.com/CDCgov/prime-reportstream/issues/12609
+                        } catch (e: IllegalArgumentException) {
+                            // fallback to string; see https://github.com/CDCgov/prime-reportstream/issues/12609
+                            if (e.cause !is NumberFormatException) throw e
+                            it
                         }
                     } else {
                         it

--- a/prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/ConstantResolver.kt
+++ b/prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/ConstantResolver.kt
@@ -164,7 +164,11 @@ class FhirPathCustomResolver(private val customFhirFunctions: FhirPathFunctions?
                 // Convert string constants that are whole integers to Integer type to facilitate math operations
                 values.map {
                     if (it is StringType && StringUtils.isNumeric(it.primitiveValue())) {
-                        IntegerType(it.primitiveValue())
+                        try {
+                            IntegerType(it.primitiveValue())
+                        } catch (e: NumberFormatException) {
+                            it // fallback to string; see https://github.com/CDCgov/prime-reportstream/issues/12609
+                        }
                     } else {
                         it
                     }

--- a/prime-router/src/test/kotlin/fhirengine/translation/hl7/utils/ConstantResolverTests.kt
+++ b/prime-router/src/test/kotlin/fhirengine/translation/hl7/utils/ConstantResolverTests.kt
@@ -136,23 +136,27 @@ class ConstantResolverTests {
     fun `test fhir path resolver multiple values`() {
         val integerValue = 99
         val stringValue = "Ninety-Nine"
+        val giantStringValue = "9999999999999999999"
 
         mockkObject(FhirPathUtils)
         every { FhirPathUtils.evaluate(any(), any(), any(), any()) } returns
-            listOf<Base>(StringType(stringValue), StringType(integerValue.toString()))
+            listOf<Base>(StringType(stringValue), StringType(integerValue.toString()), StringType(giantStringValue))
 
         val constants = sortedMapOf("const1" to "'value1'") // this does not matter but context wants something
         val context = CustomContext.addConstants(constants, CustomContext(Bundle(), Bundle()))
         val result = FhirPathCustomResolver().resolveConstant(context, "const1", false)
         assertThat(result).isNotNull()
         assertThat(result.isNotEmpty())
-        assertThat(result.size == 2)
+        assertThat(result.size == 3)
         assertThat(result[0].isPrimitive).isTrue()
         assertThat(result[0]).isInstanceOf(StringType::class.java)
         assertThat((result[0] as StringType).value).isEqualTo(stringValue)
         assertThat(result[1].isPrimitive).isTrue()
         assertThat(result[1] is IntegerType).isTrue()
         assertThat((result[1] as IntegerType).value).isEqualTo(integerValue)
+        assertThat(result[2].isPrimitive).isTrue()
+        assertThat(result[2]).isInstanceOf(StringType::class.java)
+        assertThat((result[2] as StringType).value).isEqualTo(giantStringValue)
     }
 
     @Test

